inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/dirty/sink/s3/S3DirtySink.java

```diff
@@ -0,0 +1,270 @@
+                value = FormatUtils.jsonFormat(data, labels);
+                break;
+            default:
+                throw new UnsupportedOperationException(
+                        String.format("Unsupported format for: %s", s3Options.getFormat()));
+        }
+        return value;
+    }
+
+    @Override
+    public synchronized void close() throws IOException {
+        if (!closed) {
+            closed = true;
+            if (this.scheduledFuture != null) {
+                scheduledFuture.cancel(false);
+                this.scheduler.shutdown();
+            }
+            try {
+                flush();
+            } catch (Exception e) {
+                LOG.warn("Writing records to s3 failed.", e);
+                throw new RuntimeException("Writing records to s3 failed.", e);
+            }
+        }
+    }
+
+    /**
+     * Flush data to s3
+     */
+    public synchronized void flush() {
+        flushing = true;
+        if (!hasRecords()) {
+            flushing = false;
+            return;
+        }
+        for (Entry<String, List<String>> kvs : batchMap.entrySet()) {
+            flushSingleIdentifier(kvs.getKey(), kvs.getValue());
+        }
+        batchMap.clear();
+        batchBytes = 0;
+        size = 0;
+        flushing = false;
+        LOG.info("S3 dirty sink statistics: readInNum: {}, writeOutNum: {}, errorNum: {}",
+                readInNum.get(), writeOutNum.get(), errorNum.get());
+    }
+
+    /**
+     * Flush data of single identifier to s3
+     *
+     * @param identifier The identifier of dirty data
+     * @param values The values of the identifier
+     */
+    private void flushSingleIdentifier(String identifier, List<String> values) {
+        if (values == null || values.isEmpty()) {
+            return;
+        }
+        String content = null;
+        try {
+            content = StringUtils.join(values, s3Options.getLineDelimiter());
+            s3Helper.upload(identifier, content);
+            LOG.info("Write {} records to s3 of identifier: {}", values.size(), identifier);
+            writeOutNum.addAndGet(values.size());
+            // Clean the data that has been loaded.
+            values.clear();
+        } catch (Exception e) {
+            errorNum.addAndGet(values.size());
+            if (!s3Options.ignoreSideOutputErrors()) {
+                throw new RuntimeException(
+                        String.format("Writing records to s3 of identifier:%s failed, the value: %s.",
+                                identifier, content),
+                        e);
+            }
+            LOG.warn("Writing records to s3 of identifier:{} failed and the dirty data will be throw away in the future"
+                    + " because the option 'sink.dirty.ignore-errors' is 'true'", identifier);
+        }
+    }
+
+    private boolean hasRecords() {
+        if (batchMap.isEmpty()) {
+            return false;
+        }
+        boolean hasRecords = false;
```

Comment:
can simplify to
`
        for (List<String> value : batchMap.values()) {
            if (!value.isEmpty()) {
                return true
            }
        }
        return false;
`